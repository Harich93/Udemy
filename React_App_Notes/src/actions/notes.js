
import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const stratNewNote = () => {
    return async ( dispatch, getState ) => {

        const { uid } = getState().auth;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await db.collection( `${ uid }/journal/notes` ).add( newNote )
        
        dispatch( activeNote( doc.id, newNote) );
        dispatch( addNewNote( doc.id, newNote ) );
    }

}
 
export const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote =( id, note ) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = ( uid ) => {
    return async( dispatch ) => {
        const notes = await loadNotes( uid );
        dispatch( setNotes( notes ) );
    }
}

export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes
});

export const savedNote = ( note ) => {

    return async( dispatch, getState ) => {

        const { uid } = getState().auth;
        const noteToFirestore = { ...note };
        delete noteToFirestore.id;
        // !note.url && delete note.url
        

        const arch = await db.doc( `${ uid }/journal/notes/${ note.id }` )
            .update( noteToFirestore )

        console.log( arch )

        dispatch( refreshNote( note.id, note ) );
        Swal.fire( 'Saved', note.title, 'success');
    } 
}

export const refreshNote = ( id, note ) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = ( file ) => {
    return async( dispatch, getState ) => {
        const { active: activeNote } = getState().note;
        Swal.fire({
            title: 'Uploading...',
            text: 'Please await...',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            },   
        });
        
        const fileUrl = await fileUpload( file );
        activeNote.url = fileUrl;
        
        dispatch( savedNote( activeNote ) );

        Swal.close();
    }
}

export const startDeleting = ( id ) => {
    console.log(id)
    return async( dispatch, getState ) => {
        
        const { uid } = getState().auth;
        
        await db.doc( `${ uid }/journal/notes/${ id }` ).delete()
            .then( () => Swal.fire( 'Delete', id, 'success') )
            .catch( console.log )

        dispatch( deleteNote(id) ); 
    }

}

export const deleteNote = ( id ) => ({
    type: types.notesDeleted,
    payload: id
})

export const logoutNote = () => ({
    type: types.notesLogoutCleaning
})