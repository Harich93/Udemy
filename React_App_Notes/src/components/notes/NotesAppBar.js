import moment from 'moment'
import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { savedNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {
    
    const date = moment().format( 'LL' );

    const note = useSelector(state => state.note.active );
    
    const dispatch = useDispatch();
    
    const handleSave = () => {

        dispatch( savedNote( note ) );
    }

    const handleUpdatePicture = () => {
        document.querySelector('#fileSelector').click();
    }
    
    const handleFileChanged = ( e ) => {
        const file = e.target.files[0];

        file && dispatch( startUploading( file ) );

    }

    return (
        <div className='notes__appbar'>
            <span>{ date }</span>

            <input
                id='fileSelector'
                name='file'
                type='file'
                style={ { display: 'none' } }
                onChange={ handleFileChanged }
            />

            <div>
                <button 
                    className='btn'
                    onClick={ handleUpdatePicture }    
                > 
                    Picture
                </button>

                <button 
                    className='btn'
                    onClick={ handleSave }
                > 
                    Save
                </button>
            </div>
        
        </div>
    )
}
