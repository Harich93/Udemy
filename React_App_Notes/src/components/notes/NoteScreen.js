import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, startDeleting } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const note = useSelector(state => state.note.active );
    const [ values, handleInputChange, reset ] = useForm(note);
    const { title, body, url } = values;

    const activeId = useRef( note.id );
    const activeUrl = useRef( note.url );

    const handleDelete = () => {
        dispatch( startDeleting( note.id ) );
    }
    
    useEffect(() => {

        if ( note.id !== activeId.current ) {
            reset( note );
            activeId.current = note.id;
        }
        if ( note.url !== activeUrl.current ) {
            reset( note );
            activeUrl.current = note.url;
        }

    }, [note, reset, note.url ]);

    useEffect(() => {
        
        dispatch( activeNote( values.id, { ...values } ) );

    }, [values, dispatch,] )
    
    return (
        <div className='notes___main-content'>
            <NotesAppBar />

            <div className='notes__content'>

                    <input 
                        type='text'
                        name='title'
                        placeholder='Some awasome title'
                        className='notes__title-input'
                        value={ title }
                        autoComplete='off'
                        onChange={ handleInputChange }
                    />

                    <textarea
                        placeholder='What happened today'
                        name='body'
                        className='notes__textarea'
                        value={ body }
                        onChange={ handleInputChange }
                    ></textarea>

                    {
                        url &&
                            <div className='notes__image'>
                                <img
                                    src={ url }
                                    alt='Imagen nota'
                                />
                            </div>
                    }

             
            </div>
            <button
                className='btn btn-delete'
                onClick={ handleDelete }
            >Delete</button>
        </div>
    )
}
