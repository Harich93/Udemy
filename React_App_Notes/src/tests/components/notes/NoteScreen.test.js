import '../../../setupTest';

import React from 'react';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import { NoteScreen } from '../../../components/notes/NoteScreen';
import { activeNote } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}))

const middlewrares = [thunk];
const mockStore = configureStore( middlewrares );

const initialState = {
        auth:{
            uid: '1',
            name: 'test'
        },
        ui:{
            loading: false,
            msgError: null
        },
        note: {
            active: {
                id: '123',
                title: 'Titulo',
                body: 'Body',
                url: 'http://urlImagen.jpg'
            },
            notes: []
        }
}

const store = mockStore( initialState );
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <NoteScreen />
    </Provider>
)



describe('Pruebas en <NoteScreen />', () => {
    
    test('Debe de mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();

    });

    test('Debe de llamar al activeNote', () => {
        
        wrapper.find('input[name="title"]').simulate('change',{
            target:{
                name: 'title',
                value: 'titulo cambiado'
            }
        })
        
        expect( activeNote ).toHaveBeenCalledWith('123',  {
            body: "Body",
            id: "123",
            title: "titulo cambiado",
            url: "http://urlImagen.jpg",
        })

    });
});