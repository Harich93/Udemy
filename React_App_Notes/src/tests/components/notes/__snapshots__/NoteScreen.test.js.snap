// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen /> Debe de mostrarse correctamente 1`] = `
<div
  className="notes___main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      August 4, 2021
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awasome title"
      type="text"
      value="Titulo"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happened today"
      value="Body"
    />
    <div
      className="notes__image"
    >
      <img
        alt="Imagen nota"
        src="http://urlImagen.jpg"
      />
    </div>
  </div>
  <button
    className="btn btn-delete"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
